pkg/fees/calculator.go

```diff
@@ -0,0 +1,77 @@
+package fees
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/currency"
+)
+
+type FeeCalculator struct {
+	ratesFetcher IRatesFetcher
+}
+
+func NewFeeCalculator(ratesFetcher IRatesFetcher) *FeeCalculator {
+	return &FeeCalculator{ratesFetcher: ratesFetcher}
+}
+
+func (c *FeeCalculator) CalculateBaseFee(
+	messageTime time.Time,
+	messageSize int64,
+	storageDurationDays int64,
+) (currency.PicoDollar, error) {
+	if messageSize < 0 {
+		return 0, fmt.Errorf("messageSize must be non-negative, got %d", messageSize)
+	}
+	if storageDurationDays < 0 {
+		return 0, fmt.Errorf(
+			"storageDurationDays must be non-negative, got %d",
+			storageDurationDays,
+		)
+	}
+
+	rates, err := c.ratesFetcher.GetRates(messageTime)
+	if err != nil {
+		return 0, err
+	}
+
+	// Calculate storage fee components separately to check for overflow
+	storageFeePerByte := rates.StorageFee * currency.PicoDollar(messageSize)
+	if storageFeePerByte/currency.PicoDollar(messageSize) != rates.StorageFee {
```

Comment:
Isn't this at risk of panicking because of division by zero, if `messageSize == 0`? Same for the if below if `storageDurationDays == 0`.

Reply:
well spotted!

Reply:
Great catch. I thought I went through and got all the division by 0 cases but I guess I missed one. Will fix

Reply:
Updated